include/bluetooth/mesh/msg.h

```diff
@@ -0,0 +1,226 @@
+
+/**
+ * Acknowledged message context for tracking the status of model messages pending a response.
+ */
+struct bt_mesh_msg_ack_ctx {
+	struct k_sem          sem;       /**< Sync semaphore. */
+	uint32_t              op;        /**< Opcode we're waiting for. */
+	uint16_t              dst;       /**< Address of the node that should respond. */
+	void                 *user_data; /**< User specific parameter. */
+};
+
+/** @brief Initialize an acknowledged message context.
+ *
+ *  Initializes semaphore used for synchronization between @ref bt_mesh_msg_ack_ctx_wait and
+ *  @ref bt_mesh_msg_ack_ctx_rx calls. Call this function before using @ref bt_mesh_msg_ack_ctx.
+ *
+ *  @param ack Acknowledged message context to initialize.
+ */
+static inline void bt_mesh_msg_ack_ctx_init(struct bt_mesh_msg_ack_ctx *ack)
+{
+	k_sem_init(&ack->sem, 0, 1);
+}
+
+/** @brief Reset the synchronization semaphore in an acknowledged message context.
+ *
+ *  This function aborts call to @ref bt_mesh_msg_ack_ctx_wait.
+ *
+ *  @param ack Acknowledged message context to be reset.
+ */
+static inline void bt_mesh_msg_ack_ctx_reset(struct bt_mesh_msg_ack_ctx *ack)
+{
+	k_sem_reset(&ack->sem);
+}
+
+/** @brief Clear parameters of an acknowledged message context.
+ *
+ *  This function clears the opcode, remote address and user data set
+ *  by @ref bt_mesh_msg_ack_ctx_prepare.
+ *
+ *  @param ack Acknowledged message context to be cleared.
+ */
+void bt_mesh_msg_ack_ctx_clear(struct bt_mesh_msg_ack_ctx *ack);
+
+/** @brief Prepare an acknowledged message context for the incoming message to wait.
+ *
+ *  This function sets the opcode, remote address of the incoming message and stores the user data.
+ *  Use this function before calling @ref bt_mesh_msg_ack_ctx_wait.
+ *
+ *  @param ack       Acknowledged message context to prepare.
+ *  @param op        The message OpCode.
+ *  @param dst       Destination address of the message.
+ *  @param user_data User data for the acknowledged message context.
+ *
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_msg_ack_ctx_prepare(struct bt_mesh_msg_ack_ctx *ack,
+				uint32_t op, uint16_t dst, void *user_data);
+
+/** @brief Check if the acknowledged message context is initialized with an opcode.
+ *
+ *  @param ack Acknowledged message context.
+ *
+ *  @return true if the acknowledged message context is initialized with an opcode,
+ *  false otherwise.
+ */
+static inline bool bt_mesh_msg_ack_ctx_busy(struct bt_mesh_msg_ack_ctx *ack)
+{
+	return (ack->op != 0);
+}
+
+/** @brief Wait for a message acknowledge.
+ *
+ *  This function blocks execution until @ref bt_mesh_msg_ack_ctx_rx is called or by timeout.
+ *
+ *  @param ack     Acknowledged message context of the message to wait for.
+ *  @param timeout Wait timeout in milliseconds.
+ *
+ *  @return 0 on success, or (negative) error code on failure.
+ */
+int bt_mesh_msg_ack_ctx_wait(struct bt_mesh_msg_ack_ctx *ack, int32_t timeout);
```

Comment:
Since this is being made a public API I'd change the timeout type to `k_timeout_t` and let the caller do the conversion from ms.